backend/src/router.rs

```diff
@@ -0,0 +1,504 @@
+use crate::prelude::*;
+use axum::body::HttpBody;
+use axum::routing::on;
+use axum::{handler::Handler, routing::MethodFilter, Router};
+use hyper::{Body, Method};
+use std::convert::Infallible;
+use std::marker::PhantomData;
+use std::ops::Deref;
+use thiserror::Error;
+use utoipa::openapi::PathItemType;
+use utoipa::OpenApi;
+
+#[derive(Debug, Error, PartialEq, Eq, PartialOrd, Ord)]
+pub enum RouteError {
+    #[error("No route found in OpenAPI scheme")]
+    NoRoute,
+    #[error("No method found for specified route in OpenAPI scheme")]
+    NoMethod,
+    #[error("No `operation_id` found")]
+    NoOperation,
+    #[error("OpenAPI's `operation_id` doesn't match handler's name")]
+    NoMatch,
+    #[error("Unexpected Hyper method - was it `Method::CONNECT`?")]
+    UnexpectedMethod,
+    #[error("Error occured during handler processing")]
+    InvalidHandler,
+}
+
+pub struct NoApi;
+
+impl<'a> ApiVersion<'a> for NoApi {}
+
+pub struct ApiV1;
+
+pub trait ApiVersion<'a> {
+    #[must_use]
+    fn to_path() -> &'a str {
+        ""
+    }
+}
+
+impl<'a> ApiVersion<'a> for ApiV1 {
+    fn to_path() -> &'a str {
+        "/api/v1"
+    }
+}
+
+pub struct ContextRouter<Version, Doc, S = (), B = Body> {
+    inner: Router<S, B>,
+    context: PhantomData<(Version, Doc)>,
+}
+
+impl<'a, Version, Doc, S, B> ContextRouter<Version, Doc, S, B>
+where
+    Version: ApiVersion<'a>,
+    Doc: OpenApi,
+{
+    pub fn no_context(self) -> Router<S, B> {
+        self.inner
+    }
+
+    #[must_use]
+    pub fn new_context<V, D>(self) -> ContextRouter<V, D, S, B> {
```

Comment:
That's strange. Constructor can receive `self` as an argument. Also, it is better to name it just `new`

Reply:
It's not a constructor, but more of a way to change existing context

Renamed to `change_context` and added `new` constructor